src/prefect/_internal/concurrency/timeouts.py

```diff
@@ -0,0 +1,249 @@
+
+
+@contextlib.contextmanager
+def cancel_async_after(timeout: Optional[float]):
+    """
+    Cancel any async calls within the context if it does not exit after the given
+    timeout.
+
+    A timeout error will be raised on the next `await` when the timeout expires.
+
+    Yields a `CancelContext`.
+    """
+    ctx = CancelContext(timeout=timeout)
+    if timeout is None:
+        yield ctx
+        return
+
+    try:
+        with anyio.fail_after(timeout) as cancel_scope:
+            logger.debug(
+                f"Entered asynchronous cancel context with {timeout:.2}s timeout"
+            )
+            yield ctx
+    finally:
+        ctx.cancelled = cancel_scope.cancel_called
+
+
+def get_deadline(timeout: Optional[float]):
+    """
+    Compute an deadline given a timeout.
+
+    Uses a monotonic clock.
+    """
+    if timeout is None:
+        return None
+
+    return time.monotonic() + timeout
+
+
+@contextlib.contextmanager
+def cancel_async_at(deadline: Optional[float]):
+    """
+    Cancel any async calls within the context if it does not exit by the given deadline.
+
+    Deadlines must be computed with the monotonic clock. See `get_deadline`.
+
+    A timeout error will be raised on the next `await` when the timeout expires.
+
+    Yields a `CancelContext`.
+    """
+    if deadline is None:
+        yield CancelContext(timeout=None)
+        return
+
+    timeout = max(0, deadline - time.monotonic())
+
+    ctx = CancelContext(timeout=timeout)
+    try:
+        with cancel_async_after(timeout) as inner_ctx:
+            yield ctx
+    finally:
+        ctx.cancelled = inner_ctx.cancelled
+
+
+@contextlib.contextmanager
+def cancel_sync_at(deadline: Optional[float]):
+    """
+    Cancel any sync calls within the context if it does not exit by the given deadline.
+
+    Deadlines must be computed with the monotonic clock. See `get_deadline`.
+
+    The cancel method varies depending on if this is called in the main thread or not.
+    See `cancel_sync_after` for details
+
+    Yields a `CancelContext`.
+    """
+    if deadline is None:
+        yield CancelContext(timeout=None)
+        return
+
+    timeout = max(0, deadline - time.monotonic())
+
+    ctx = CancelContext(timeout=timeout)
+    try:
+        with cancel_sync_after(timeout) as inner_ctx:
+            yield ctx
+    finally:
+        ctx.cancelled = inner_ctx.cancelled
+
+
+@contextlib.contextmanager
+def cancel_sync_after(timeout: Optional[float]):
+    """
+    Cancel any sync calls within the context if it does not exit after the given
+    timeout.
+
+    The timeout method varies depending on if this is called in the main thread or not.
+    See `_alarm_based_timeout` and `_watcher_thread_based_timeout` for details.
+
+    Yields a `CancelContext`.
+    """
+    ctx = CancelContext(timeout=timeout)
+    if timeout is None:
+        yield ctx
+        return
+
+    if sys.platform.startswith("win"):
+        # Timeouts cannot be enforced on Windows
```

Comment:
Maybe a `logger.warning` to that effect here?

Reply:
Seems wise in the meantime, I'm guessing we can have some sort of support eventually.